torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,392 @@
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(symbolically_differentiable) {}
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(isDifferentiable(*this->graph)) {}
+
+  // entry point where execution begins
+  variable_tensor_list run(variable_tensor_list inputs) {
+    // this is the fallback pathway, when we cannot differentiate
+    if(!optimize || (!symbolically_differentiable && needsGradient(inputs))) {
+      auto & fb = getOrCreateAutogradFallback();
+      InterpreterState state(fb);
+      tensor_list outputs;
+      state.runOneStage(std::move(inputs), outputs);
+      // note: we never unwrapped inputs, because we want autograd to record the trace
+      return variable_tensor_list(std::move(outputs));
+    }
+
+    // either we can symbolically differentiate, or we do not need a gradient.
+    // go down the route where we treat the inputs as tensors
+    // and fully optimize
+    auto & implementation = getOrCompile(inputs);
+    return implementation.run(std::move(inputs));
+  }
+
+private:
+
+  static bool needsGradient(const variable_tensor_list & inputs) {
+    if (!autograd::GradMode::is_enabled()) {
+      return false;
+    }
+    for (const auto & tensor : inputs) {
+      if(tensor.defined() && static_cast<const Variable&>(tensor).requires_grad())
+        return true;
+    }
+    return false;
+  }
+  static bool isDifferentiable(Graph & g) {
+    for(auto n : g.nodes()) {
+      if(!jit::isDifferentiable(n))
+        return false;
+    }
+    return true;
+  }
+
+  void runOptimization(std::shared_ptr<Graph> & graph, bool graphMustSupportVariables) {
+
+    // these optimizations must run in the presence of variables
+    // and when shape information is not statically known.
+
+    EliminateDeadCode(graph);
+    CheckInplace(graph);
+    EliminateCommonSubexpression(graph);
+
+    if (!graphMustSupportVariables) {
+      // These optimizations can introduce operators like FusionGroup that
+      // do not work on variables
+
+      // They also may assume that concrete sizes/strides are availiable
+
+      //TODO: create peephole optimizations that are safe to run
+      // when we are using variables, and when we do not know sizes.
+      PeepholeOptimize(graph);
+      // TODO: remove mandatory size checking in BatchMM, otherwise
+      // it works fine on variables.
+      BatchMM(graph);
+      FuseGraph(graph);
+    }
+  }
+  Code & getOrCreateAutogradFallback() {
+    std::lock_guard<std::mutex> lock(compile_mutex);
+    if(autograd_fallback) {
+      return autograd_fallback;
+    }
+    auto graph_ = graph->copy();
+    if(optimize) {
+      CreateAutodiffSubgraphs(*graph_);
+      runOptimization(graph_, /*graphMustSupportVariables=*/true);
+    }
+    autograd_fallback = Code(graph_);
+    return autograd_fallback;
```

Comment:
Uhhhh... are you sure you didn't want to return a const reference here? Otherwise it seems very easy to break the locking convention from calling code.